File: tests/test_classifier.py
"""Tests for the rule-based classifier against the public sample cases."""

from app.classifier import classify
from app.schemas import CaseType, Severity, Department


def assert_case(msg: str, *, case: CaseType, severity: Severity | None = None):
    out = classify(msg)
    assert out["case_type"] == case, f"for {msg!r} expected {case} got {out['case_type']} (signals={out['signals']})"
    if severity is not None:
        assert out["severity"] == severity, f"for {msg!r} expected severity {severity} got {out['severity']}"
    assert 0.0 <= out["confidence"] <= 1.0
    assert out["agent_summary"]
    # Summary must be a single string, no list/dict types.
    assert isinstance(out["agent_summary"], str)


# --- Public sample cases from the spec --------------------------------------

def test_sample_1_wrong_transfer():
    assert_case(
        "I sent 3000 to wrong number",
        case=CaseType.WRONG_TRANSFER,
        severity=Severity.HIGH,
    )

def test_sample_2_payment_failed():
    assert_case(
        "Payment failed but balance deducted",
        case=CaseType.PAYMENT_FAILED,
        severity=Severity.HIGH,
    )

def test_sample_3_phishing_otp_call():
    out = classify("Someone called asking my OTP, is that bKash?")
    assert out["case_type"] == CaseType.PHISHING
    assert out["severity"] == Severity.CRITICAL
    assert out["department"] == Department.FRAUD_RISK
    assert out["human_review_required"] is True

def test_sample_4_refund_changed_mind():
    assert_case(
        "Please refund my last transaction, I changed my mind",
        case=CaseType.REFUND_REQUEST,
        severity=Severity.LOW,
    )

def test_sample_5_app_crash_other():
    assert_case(
        "App crashed when I opened it",
        case=CaseType.OTHER,
    )


# --- Extra coverage: phishing variants, refunds, edge cases ----------------

def test_phishing_pin_ask():
    out = classify("An officer is calling and asking me to share my PIN to verify my bKash account.")
    assert out["case_type"] == CaseType.PHISHING
    assert out["human_review_required"] is True

def test_phishing_link():
    out = classify("I got an SMS saying my account will be blocked. Click this link http://bit.ly/x to verify.")
    assert out["case_type"] == CaseType.PHISHING

def test_refund_duplicate_charge():
    out = classify("I was charged twice for the same order, please refund the duplicate payment.")
    assert out["case_type"] == CaseType.REFUND_REQUEST

def test_refund_merchant():
    out = classify("I want to return my order to the merchant and get a refund.")
    assert out["case_type"] == CaseType.REFUND_REQUEST

def test_other_unrelated():
    out = classify("Hello, what are your business hours?")
    assert out["case_type"] == CaseType.OTHER

def test_summary_mentions_amount_when_present():
    out = classify("I sent 5000 taka to a wrong number, please help.")
    assert "5000" in out["agent_summary"]


# --- Severity rules --------------------------------------------------------

def test_phishing_is_critical():
    out = classify("Someone asked for my OTP on call.")
    assert out["severity"] == Severity.CRITICAL

def test_refund_low_when_changed_mind():
    out = classify("Please cancel and refund my last transaction, I changed my mind.")
    assert out["severity"] == Severity.LOW


# --- Human review rule -----------------------------------------------------

def test_human_review_for_phishing():
    assert classify("He is asking my OTP.")["human_review_required"] is True

def test_human_review_for_high_severity():
    # High severity alone does NOT require human review per the spec.
    out = classify("I sent 3000 to wrong number")
    assert out["severity"] == Severity.HIGH
    assert out["human_review_required"] is False

File: tests/test_classifier.js
'use strict';

const test = require('node:test');
const assert = require('node:assert/strict');
const { classify } = require('../server.js');

test('classifier: wrong_transfer', () => {
  const r = classify('I sent 5000 taka to a wrong number this morning, please help me get it back');
  assert.equal(r.case_type, 'wrong_transfer');
  assert.equal(r.severity, 'high');
  assert.equal(r.department, 'dispute_resolution');
  assert.equal(r.human_review_required, false);
});

test('classifier: payment_failed', () => {
  const r = classify('Payment failed but balance deducted. Please check.');
  assert.equal(r.case_type, 'payment_failed');
  assert.equal(r.severity, 'high');
  assert.equal(r.department, 'payments_ops');
});

test('classifier: phishing_or_social_engineering', () => {
  const r = classify('Someone called asking my OTP, is that bKash?');
  assert.equal(r.case_type, 'phishing_or_social_engineering');
  assert.equal(r.severity, 'critical');
  assert.equal(r.department, 'fraud_risk');
  assert.equal(r.human_review_required, true);
});

test('classifier: refund_request (low - changed my mind)', () => {
  const r = classify('Please refund my last transaction, I changed my mind.');
  assert.equal(r.case_type, 'refund_request');
  assert.equal(r.severity, 'low');
  assert.equal(r.department, 'dispute_resolution');
});

test('classifier: refund_request (medium - generic refund)', () => {
  const r = classify('Please refund my last transaction, the merchant refused to take it back.');
  assert.equal(r.case_type, 'refund_request');
  assert.equal(r.severity, 'medium');
});

test('classifier: other', () => {
  const r = classify('App crashed when I opened it.');
  assert.equal(r.case_type, 'other');
  assert.equal(r.department, 'customer_support');
});

test('classifier: phishing overrides even with refund keywords present', () => {
  const r = classify('Please refund me, but first share your OTP with us.');
  assert.equal(r.case_type, 'phishing_or_social_engineering');
  assert.equal(r.human_review_required, true);
});

test('classifier: human_review_required flips on critical severity', () => {
  const r = classify('Someone saying they are bKash officer is asking for my PIN to verify account.');
  assert.equal(r.severity, 'critical');
  assert.equal(r.human_review_required, true);
});

test('classifier: confidence is in [0, 1]', () => {
  const samples = [
    'I sent 5000 taka to a wrong number',
    'Payment failed but balance deducted',
    'Someone called asking my OTP',
    'Please refund my last transaction, I changed my mind.',
    'App crashed when I opened it.',
  ];
  for (const m of samples) {
    const r = classify(m);
    assert.ok(r.confidence >= 0 && r.confidence <= 1, `confidence ${r.confidence} out of range`);
  }
});

test('classifier: every response has all required fields', () => {
  const r = classify('I sent 5000 taka to a wrong number');
  for (const k of ['case_type','severity','department','agent_summary','human_review_required','confidence','signals']) {
    assert.ok(k in r, `missing field ${k}`);
  }
});